### ISSUE TITLE
NgbOffcanvasRef and NgbModalRef should expose a ComponentRef

### ISSUE THREAD

**Issue by pkfms** [ng-bootstrap/ng-bootstrap#4688]
### Bug description:
NgbOffcanvasRef and NgbModalRef should expose a ComponentRef when using a component type as content.
Currently the NgbOffcanvasRef exposes a componentInstance. This makes it hard to use components with Signal inputs, as you cannot do myComponentInstance.myProp = value. Instead, the proper way is to use the function setInput on a componentRef. setInput also triggers change detection and ngOnChanges correctly.

### Link to minimally-working StackBlitz that reproduces the issue:
https://stackblitz.com/edit/angular-5zsark-l3yj9i

### Versions of Angular, ng-bootstrap and Bootstrap:

Angular: 17.3.1

ng-bootstrap: 16.0.0

Bootstrap: 5.3.3

**Comment by jnizet** [comment]
I agree that it's not clean setting inputs programmatically on the modal component. OTOH, I've always wondered why the examples in the documentation decorate the properties set by the opening component with `@Input()`. They're not inputs and have no reason to be. 

I generally expose an `initialize()` method in the modal component that the opening component calls once the component has been created by the modal service.

In your stackblitz, you could replace the inputs by regular writable signals.

**Comment by pkfms** [comment]
Sure, that works in this example. I'm trying to share the content components and do things like opening a popover on desktop and offcanvas on mobile. or modal/offcanvas

**Comment by GaryChamberlain** [comment]
I have the same problem. I hacked it for now. Hopefully the componentRef will be exposed soon so I can clean this up.

`/**`
` * TODO: This is a hack to get the ComponentRef because NgbModalRef does not expose it.`
` * See https://github.com/ng-bootstrap/ng-bootstrap/issues/4688`
` */`
`const componentRef: ComponentRef<C> = ref['_contentRef'].componentRef;`
`componentRef.setInput('data', data);`

**Comment by screambeard** [comment]
Agree that this would be a great addition to the modal API interface.

**Comment by mkreidenweis-schulmngr** [comment]
Duplicate of #4664 ?

**Comment by Kludgy4** [comment]
I have also encountered this issue. This would be a great solution

**Comment by majora2007** [comment]
Any update from the team on this? We really need a way to set signals when opening programmatically.

**Comment by bougwal** [comment]
Q: any updates on this outstanding demand ?

**Comment by majora2007** [comment]
I ended up rolling my own wrappers:
https://github.com/Kareadita/Kavita/blob/develop/UI/Web/src/app/_services/modal.service.ts
https://github.com/Kareadita/Kavita/blob/develop/UI/Web/src/app/_services/drawer.service.ts

Modal service looks a bit scary, but it was necessary to be able to get good type sense for the properties.
